File: CopyListWithRandomPtr.java

// Definition for a Node.

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

// O(n) time O(n) space
// class Solution {
// public Node copyRandomList(Node head) {
// if (head == null)
// return null;
// HashMap<Node, Node> map = new HashMap<>();
// Node copyHead = new Node(head.val);
// map.put(head, copyHead);
// Node curr = head;
// Node cpcurr = copyHead;
// while (curr != null) {
// cpcurr.next = clone(curr.next, map);
// cpcurr.random = clone(curr.random, map);
// curr = curr.next;
// cpcurr = cpcurr.next;
// }
// return copyHead;
// }

// private Node clone(Node node, HashMap<Node, Node> map) {
// if (node == null)
// return null;
// if (!map.containsKey(node)) {
// Node newNode = new Node(node.val);
// map.put(node, newNode);
// }
// return map.get(node);
// }
// }

// O(n) time O(1) space
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null)
            return null;
        Node curr = head;
        while (curr != null) {
            Node newNode = new Node(curr.val);
            newNode.next = curr.next;
            curr.next = newNode;
            curr = curr.next.next;
        }
        curr = head;
        Node copyCurr = head.next;
        Node copyHead = head.next;
        while (curr != null) {
            if (curr.random != null) {
                copyCurr.random = curr.random.next;
            }
            curr = curr.next.next;
            if (copyCurr.next != null)
                copyCurr = copyCurr.next.next;
        }
        curr = head;
        copyCurr = head.next;
        while (curr != null) {
            curr.next = curr.next.next;
            if (copyCurr.next != null)
                copyCurr.next = copyCurr.next.next;
            curr = curr.next;
            copyCurr = copyCurr.next;
        }
        return copyHead;
    }
}

File: CopyListWithRandomPtr.py
'''
    Time Complexity:
        O(n) (where n = number of nodes)

    Space Complexity:
        O(n) (where n = number of nodes)

    Did this code successfully run on LeetCode?:
        Yes

    Problems faced while coding this:
        None

    Approach:
        Keep a mapping of {old_node: new_node} in a HashMap.
'''

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def __init__(self):
        self.hash_map = {}

    def copyRandomList(self, head: 'Node') -> 'Node':
        return self.deep_copy(head)

    def deep_copy(self, node):
        if node is None:
            return None

        if node in self.hash_map:
            return self.hash_map[node]

        n = Node(node.val)
        self.hash_map[node] = n
        n.next = self.deep_copy(node.next)
        n.random = self.deep_copy(node.random)
        return n
